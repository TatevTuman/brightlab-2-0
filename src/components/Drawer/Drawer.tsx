import React, { memo } from 'react'
import { createPortal } from 'react-dom'
import { Children } from '@types'
import { useDrawer } from '@hooks'
import './Drawer.scss'

export interface DrawerProps {
  drawerName: string
  children: Children
}

type DrawerType = React.FC<DrawerProps> & {}

const Drawer: DrawerType = props => {
  const { drawerName, children } = props
  const { drawer: isDrawerOpened, closeDrawer } = useDrawer(drawerName)
  const container = document.getElementById('drawers')

  if (!isDrawerOpened) return null

  const DrawerComponent = memo(() => {
    return (
      <div id={drawerName} className={'drawer'} role={'complementary'}>
        <div className={'drawer-content'}>{children}</div>
      </div>
    )
  })

  DrawerComponent.displayName = drawerName

  return createPortal(<DrawerComponent />, container!)
}

export default Drawer
