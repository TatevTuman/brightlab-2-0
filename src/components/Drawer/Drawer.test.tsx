import React from 'react'
import { render, cleanup, waitFor, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Button } from '@elements'
import { useDrawer } from '@hooks'
import Drawer from './Drawer'

jest.useFakeTimers()

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

const TestDrawer = () => {
  const { openDrawer, closeDrawer } = useDrawer('TestDrawer')

  return (
    <>
      <div id={'drawers'} data-testid={'drawers'} />
      <Drawer drawerName={'TestDrawer'}>
        <div data-testid={'drawer'}>TestDrawer</div>
      </Drawer>
      <Button onClick={openDrawer}>Open TestDrawer</Button>
      <Button onClick={closeDrawer}>Close TestDrawer</Button>
    </>
  )
}

describe('Drawer', () => {
  it('renders correctly', async () => {
    const { container, getByText, getByTestId } = render(<TestDrawer />)
    const awaitedContainer = await waitFor(() => container)
    const drawersContainer = getByTestId('drawers')

    let drawer = awaitedContainer.querySelector('div[data-testid="drawer"]')

    expect(drawer).toBeNull()
    expect(drawersContainer.childElementCount).toBe(0)

    const openButton = getByText('Open TestDrawer')
    userEvent.click(openButton)

    drawer = awaitedContainer.querySelector('div[data-testid="drawer"]')

    expect(drawer).toBeInTheDocument()
    expect(drawersContainer.childElementCount).toBe(1)
    expect(awaitedContainer).toMatchSnapshot()

    const closeButton = getByText('Close TestDrawer')
    userEvent.click(closeButton)

    setTimeout(() => {
      drawer = awaitedContainer.querySelector('div[data-testid="drawer"]')

      expect(drawer).toBeNull()
      expect(drawersContainer.childElementCount).toBe(0)
    }, 1000)

    act(() => {
      jest.runAllTimers()
    })
  })
})
