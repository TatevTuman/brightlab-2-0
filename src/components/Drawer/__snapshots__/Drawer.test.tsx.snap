// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer renders correctly 1`] = `
<div>
  <div
    data-testid="drawers"
    id="drawers"
  >
    <div
      class="drawer"
      id="TestDrawer"
      role="complementary"
    >
      <div
        class="drawer-content"
      >
        <div
          data-testid="drawer"
        >
          TestDrawer
        </div>
      </div>
    </div>
  </div>
  <button
    class="button button__primary-lg "
    data-centered="false"
    data-disabled="false"
    type="button"
  >
    Open TestDrawer
  </button>
  <button
    class="button button__primary-lg "
    data-centered="false"
    data-disabled="false"
    type="button"
  >
    Close TestDrawer
  </button>
</div>
`;
